src/multiplex/multiplex.rs

```diff
@@ -377,75 +451,175 @@ impl<S, T, E> Multiplex<S, T>
-            } else {
-                break;
+            match try!(self.dispatch.poll()) {
+                Async::Ready(Some((id, Ok(message)))) => {
+                    self.dispatch_made_progress();
+
+                    trace!("   --> got message");
+                    try!(self.write_in_message(id, message));
+                }
+                Async::Ready(Some((id, Err(error)))) => {
+                    self.dispatch_made_progress();
+
+                    trace!("   --> got error");
+                    try!(self.write_in_error(id, error));
+                }
+                Async::Ready(None) => {
+                    trace!("   --> got error");
+                    trace!("   --> got None");
+                    // The service is done with the connection. In this case, a
+                    // `Done` frame should be written to the transport and the
+                    // transport should start shutting down.
+                    //
+                    // However, the `Done` frame should only be written once
+                    // all the in-flight bodies have been written.
+                    //
+                    // For now, do nothing...
+                    break;
+                }
+                // Nothing to dispatch
+                Async::NotReady => break,
             }
         }
 
         trace!("   --> transport not ready");
+        self.blocked_on_flush.transport_not_write_ready();
 
         Ok(())
     }
 
-    fn write_in_message(&mut self, id: RequestId, message: Result<Message<S::InMsg, S::InBodyStream>, S::Error>) -> io::Result<()> {
-        match message {
-            Ok(Message::WithoutBody(val)) => {
-                trace!("got in_flight value without body");
-                try!(self.transport.write(Frame::Message(id, val)));
+    fn write_in_message(&mut self,
+                        id: RequestId,
+                        message: Message<T::In, T::Stream>)
+                        -> io::Result<()>
+    {
+        let (message, body) = match message {
+            Message::WithBody(message, rx) => (message, Some(rx)),
+            Message::WithoutBody(message) => (message, None),
+        };
+
+        // Create the frame
+        let frame = Frame::Message {
+            id: id,
+            message: message,
+            body: body.is_some()
+        };
+
+        // Write the frame
+        try!(self.dispatch.transport().write(frame));
+        self.blocked_on_flush.wrote_frame();
+
+        match self.exchanges.entry(id) {
+            Entry::Occupied(mut e) => {
+                assert!(!e.get().responded, "invalid exchange state");
+                assert!(e.get().is_outbound());
+
+                // Track that the exchange has been responded to
+                e.get_mut().responded = true;
+
+                // Set the body receiver
+                e.get_mut().in_body = body;
+
+                // If the exchange is complete, clean up the resources
```

Comment:
This was initially surprising to me, but perhaps this comment could be expanded to indicate that if `body` is `None` and we already read the data, everything is done?